### PR TITLE
Warning message for lvm devmapper running on top of loopback devices

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan shishir.mahajan@redhat.com

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
There are 2 changes in this PR:

1) We want to warn users if they are running devmapper on loopback devices. This is not production ready and should be discouraged.
2) Users should have the ability to suppress this warning, and not constantly pestered by it.

Docs are updated for more elaborate description.

**Comment by thaJeztah** [comment]
ping @vbatts

**Comment by vbatts** [comment]
it doesn't say expressly say "it is not supported", but hopefully will once we decide to never fall back to devicemapper on loop-lvm, because upstream lvm absolutely *does* *not* support this loop-lvm setup. 

LGTM

**Comment by rhvgoyal** [comment]
This patch takes warning all the way to client. Assumption here is that not many people might bother to read syslogs and notice warning.

LGTM.

**Comment by shishir-a412ed** [comment]
Following code and doc changes are taken from PR #12404 

1) daemon/graphdriver/devmapper/README.md: doc change.
2) lvm devmapper on loopback devices check: daemon/graphdriver/devmapper/deviceset.go

File: api/client/run.go [hunk 1]
```diff
@@ -188,10 +188,15 @@ func (cli *DockerCli) CmdRun(args ...string) error {
 	}()
 
 	//start the container
-	if _, _, err = readBody(cli.call("POST", "/containers/"+createResponse.ID+"/start", nil, nil)); err != nil {
+	var obj []byte
+	if obj, _, err = readBody(cli.call("POST", "/containers/"+createResponse.ID+"/start", nil, nil)); err != nil {
```
Comment (runcom) [inline api/client/run.go]: nit but can we avoid the var declaration and split in `obj, _, err := ...` and `if err != nil {}` otherwise `obj`?

**Comment by crosbymichael** [comment]
Why should this be returned to the client?  It's a daemon option and the client should not know about this.  The warning should be to the admin that configured and started docker and we should not have a specific flag to stop this warning.

**Comment by vbatts** [comment]
@crosbymichael while I understand what you're saying, the fact that devicemapper loop-lvm is what so many distros allow users to default to, the fact is that this not an admin specifically configuring to use it. They install docker, and get an awful experience with devicemapper on loop-lvm, and have no idea why. We could hope for users that intentionally configure the storage driver for their optimal use case ...

**Comment by cgwalters** [comment]
Can we land https://github.com/docker/docker/pull/13323 first, then rebase this on top?

**Comment by cgwalters** [comment]
One issue with this is that for any usage of the client binary in scripts, all of a sudden we're injecting new output to stderr.

```
# docker run -ti centos echo hello |& wc -l
2
```

I hope not too many people are parsing stderr, but it's something to be aware of.

One compromise we might make is to only emit the warning when the client is connected to a tty.

**Comment by icecrime** [comment]
Agree with Michael: error message shouldn't be shown on the client side (and especially not mixed with regular output).

File: api/server/server.go [hunk 2]
```diff
@@ -1047,6 +1048,11 @@ func (s *Server) postContainersStart(version version.Version, w http.ResponseWri
 		}
 		return err
 	}
+
+	if devmapper.WarnOnLoopback && devmapper.LoopbackInUse {
+		fmt.Fprintln(w, "Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.")
```
Comment (icecrime) [inline api/server/server.go]: This should be a log.

Comment (rhatdan) [inline api/server/server.go]: I am not a big fan of this either, the problem is our file system team is wants it badly to make sure customers do not accidentally  put this into production, and they feel that the log message will be ignored.  I believe there is another patch that writes this message to the syslog system.

**Comment by shishir-a412ed** [comment]
seems like we are going in an impasse :)
We understand that warning message on client side may not be the best place, but do we really want docker customers to be unhappy with the experience they get running docker ?

Scenario: I am a docker user, not a storage guy. I just started docker (which is running loopback devices by default). I didn't set any options. Now my docker is running slow. It didn't blew up, it's just running slow. I have no clue why but I am not happy with the experience !

Honestly, This is not the first time we are having a warning message on client side. Try updating an image with `docker load` command. you ll get a warning message that the old image is getting renamed to none:none. The load will still go through as it's a warning message. I think this one is much more critical.

**Comment by mrjana** [comment]
@shishir-a412ed Based on the discussion I and @tiborvass had the warning definitely needs to be a part of daemon log and it can also be be part of `docker info` so users can see that straight away when they either see the syslog or `docker info`. Showing it on the client side is tricky because the user cannot distinguish if the message is coming from the daemon or from the application running in the container unless the container was started in detached mode.

**Comment by rhatdan** [comment]
I agree this should be put into two different pull requests.  One to reveal the information in the docker logs, and docker info.  And one to output the data to the client.  If docker only accepted the first patch, I would be fine with it.  We are also working on a tool to help users to migrate off of one back end to another, this would allow someone who went into production using loopback a way to migrate off without loosing their containers.  It should also allow people to move from one back end to another. Say try out Overlayfs.

**Comment by rhvgoyal** [comment]
docker info has the information present that we are using loop devices. 

Putting a warning in there too, feels little odd to me. I feel warning should be part of syslogs and if possible carry it all the way to the client.

**Comment by rhatdan** [comment]
I look at that data and I have no idea what it means,   Adding (Developer Mode) or (No Production Mode) would be useful.

**Comment by rhvgoyal** [comment]
I am afraid then this mode field will start getting overloaded soon and one will have to go through logs anyway to figure out why their docker is in developer mode and not in production mode. And if one has to go through logs anyway, why to introduce more fields in docker info.

**Comment by rhvgoyal** [comment]
And then their will be arguments about when something is considered developer mode and when does it go into production mode. For example.

- Using overlay is developer mode or production mode.
- Say there is some configuration of lvm thin pool which is not recommended and user has setup. Then should we transition docker into developer mode?

IOW, what I am trying to say is that trying to classify docker/storage configuration into "Developer/Production" mode is not going to be sufficient. There can be many reasons for docker to be in certain mode and user will have to walk through the logs anyway. It is not a replacement for that.

**Comment by icecrime** [comment]
@shishir-a412ed Some concerns have been raised some time ago by @crosbymichael and @mrjana (with @tiborvass), especially regarding the fact that this error is produced to the client, as part of a `docker run`. Can you please address them? Thanks.

**Comment by shishir-a412ed** [comment]
@rhvgoyal @rhatdan
What do you guys wanna do ? If you guys are okay with putting the message on the daemon log, I can quickly make the changes. 

If not, I can just close this PR. I see a lot of back and forth happening on this one, with no consensus.

**Comment by rhatdan** [comment]
I think we should just log to the syslog upstream.  Our file system teams insist on us logging to the client, so we need to carry the patch until we can change the default.

**Comment by rhvgoyal** [comment]
Agreed. One step at a time. Let us first log it in syslog and then see what else can be done to make it more explicit.

**Comment by shishir-a412ed** [comment]
@icecrime I have made the changes.

File: daemon/graphdriver/devmapper/deviceset.go [hunk 3]
```diff
@@ -1397,6 +1397,12 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
 		}
 	}
 
+	if devices.thinPoolDevice == "" {
+		if devices.metadataLoopFile != "" || devices.dataLoopFile != "" {
+			logrus.Warnf("Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev`.")
```
Comment (tiborvass) [inline daemon/graphdriver/devmapper/deviceset.go]: would be great to output a URL to the docs where users can read more about how to properly configure devmapper

Comment (shishir-a412ed) [inline daemon/graphdriver/devmapper/deviceset.go]: @tiborvass I have made some changes. Take a look now and see if this is good.

**Comment by cpuguy83** [comment]
LGTM

**Comment by calavera** [comment]
LGTM. Moving to docs review

**Comment by bfirsh** [comment]
@moxiegirl @SvenDowideit Look good?

**Comment by SvenDowideit** [comment]
LGTM - @thaJeztah @jamtur01

File: daemon/graphdriver/devmapper/README.md [hunk 4]
```diff
@@ -3,12 +3,21 @@
...
-one for metadata.  By default these block devices are created
-automatically by using loopback mounts of automatically created sparse
-files.
+module (dm-thinp) to implement CoW snapshots.  The preferred model is
+to have a thin pool reserved outside of Docker, and passed to the
+daemon via the `--storage-opt dm.thinpooldev` option.
+
+As a fallback if no thin pool is provided, loopback files will be
+created.  Loopback is very slow, but can be used without any
+pre-configuration of storage.  It is *strongly recommended* to not use
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: This phraseology is very awkward - perhaps: "It is *strongly recommended* that you do not use loopback in production." Ditto below.

File: daemon/graphdriver/devmapper/README.md [hunk 5]
```diff
@@ -3,12 +3,21 @@
...
+loopback in production.  Ensure your docker daemon has a
+`--storage-opt dm.thinpooldev` argument provided.
+
+In loopback, each devicemapper graph location (typically
+`/var/lib/docker/devicemapper`, $graph below) a thin pool is created
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: The $graph reference isn't clear - I'd rewrite this sentence.

File: docs/reference/commandline/daemon.md [hunk 6]
```diff
@@ -192,6 +192,12 @@ options for `zfs` start with `zfs`.
      resize support, dynamically changing thin-pool features, automatic thinp
      metadata checking when lvm activates the thin-pool, etc.
 
+     As a fallback if no thin pool is provided, loopback files will be
+     created.  Loopback is very slow, but can be used without any
+     pre-configuration of storage.  It is *strongly recommended* to not use
+     loopback in production.  Ensure your docker daemon has a
```
Comment (jamtur01) [inline docs/reference/commandline/daemon.md]: Docker

File: daemon/graphdriver/devmapper/README.md [hunk 7]
```diff
@@ -3,12 +3,21 @@
...
-automatically by using loopback mounts of automatically created sparse
-files.
+module (dm-thinp) to implement CoW snapshots.  The preferred model is
+to have a thin pool reserved outside of Docker, and passed to the
+daemon via the `--storage-opt dm.thinpooldev` option.
+
+As a fallback if no thin pool is provided, loopback files will be
+created.  Loopback is very slow, but can be used without any
+pre-configuration of storage.  It is *strongly recommended* to not use
+loopback in production.  Ensure your docker daemon has a
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: Docker

**Comment by shishir-a412ed** [comment]
@jamtur01 Please check now.

File: daemon/graphdriver/devmapper/README.md [hunk 8]
```diff
@@ -3,22 +3,33 @@
 ### Theory of operation
 
 The device mapper graphdriver uses the device mapper thin provisioning
-module (dm-thinp) to implement CoW snapshots. For each devicemapper
-graph location (typically `/var/lib/docker/devicemapper`, $graph below)
-a thin pool is created based on two block devices, one for data and
-one for metadata.  By default these block devices are created
-automatically by using loopback mounts of automatically created sparse
+module (dm-thinp) to implement CoW snapshots.  The preferred model is
+to have a thin pool reserved outside of Docker, and passed to the
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: You don't need a comma here after Docker.

**Comment by jamtur01** [comment]
Docs broadly LGTM

File: daemon/graphdriver/devmapper/README.md (hunk 8 above)
Comment (shishir-a412ed) [inline daemon/graphdriver/devmapper/README.md]: @jamtur01 Fixed.

**Comment by thaJeztah** [comment]
docs LGTM, thanks @shishir-a412ed!

**Pull request merged**